Regression Team TS1 Notebook

# Problem Statement

The task is to create a model that can accurately predict what the price of Golden Delicious Apples per kilogram.

## Define the objective in business terms.

Selecting the correct price of the apples is vital to the business's success.  If the apples price is too low, then the business will not turn a profit, but if the price is too high, then the apples wont sell, the company will be left with too much stock and revenue will suffer.

## How the solution will be implemented

The model will allow the business to make informed decisions on what pricing strategy needs to be used.

## How will this problem be solved?

* Supervised learning using multiple linear regression.



## How should performance be measured?

Perfomance will be measurred using test data.

## Is the performance measure aligned with the business objective?

## What would be the minimum performance needed to reach the business objective?

## What are comparable problems? Can you reuse experience or tools?

## Is human expertise available?

## How would you solve the problem manually?

## List the assumptions you (or others) have made so far.

## Verify the assumptions if possible.

# Importing the data and packages

Both train and test data is given

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

In [5]:
dftest = pd.read_csv('df-test_set.csv')
dftrain = pd.read_csv('df-train_set.csv')

# Explore the data

In [3]:
dftest.head()

Index              Province Container Size_Grade  Weight_Kg  \
0      1  W.CAPE-BERGRIVER ETC     EC120         1M       12.0   
1      2  W.CAPE-BERGRIVER ETC     M4183         1X       18.3   
2      3  W.CAPE-BERGRIVER ETC     EC120         1S       12.0   
3      4  W.CAPE-BERGRIVER ETC     M4183         1M       18.3   
4      5  W.CAPE-BERGRIVER ETC     M4183         1L       18.3   

              Commodities        Date  Low_Price  High_Price  Sales_Total  \
0  APPLE GOLDEN DELICIOUS  2020-07-09      128.0       136.0       5008.0   
1  APPLE GOLDEN DELICIOUS  2020-01-20      220.0       220.0       1760.0   
2  APPLE GOLDEN DELICIOUS  2020-08-19      120.0       120.0        720.0   
3  APPLE GOLDEN DELICIOUS  2020-05-06      160.0       160.0        160.0   
4  APPLE GOLDEN DELICIOUS  2020-05-04      140.0       160.0      14140.0   

   Total_Qty_Sold  Total_Kg_Sold  Stock_On_Hand  
0              38          456.0              0  
1               8          146.4              2  
2               6           72.0             45  
3               1           18.3              8  
4             100         1830.0             19

In [13]:
dftrain.head()

Province Container Size_Grade  Weight_Kg  \
64371              TRANSVAAL     AG100         1M       10.0   
64372  IMPORTS - OUTSIDE RSA     BC180          X       18.0   
64373              TRANSVAAL     TS035         1M        3.5   
64374                  NATAL     BJ100         2M       10.0   
64375                   CAPE     M4183         1L       18.3   

              Commodities        Date  Low_Price  High_Price  Sales_Total  \
64371         ONIONS MILD  2020-05-18       70.0        75.0      19695.0   
64372              BANANA  2020-02-05      160.0       220.0      13560.0   
64373  TOMATOES-LONG LIFE  2020-08-20       20.0        20.0       6000.0   
64374               LEMON  2020-08-18       50.0        80.0      16730.0   
64375          APPLE FUJI  2020-08-06      150.0       150.0        150.0   

       Total_Qty_Sold  Total_Kg_Sold  Stock_On_Hand  avg_price_per_kg  
64371             281         2810.0           3583              7.01  
64372              72         1296.0            887             10.46  
64373             300         1050.0            900              5.71  
64374             277         2770.0            372              6.04  
64375               1           18.3              1              8.20

## Study each attribute and its characteristics.

Looking at the top five rows of our data we can see all our features as well as the types of data we are working with.

Our features are:

    - Province
    - Container
    - Size_grade
    - Weight_kg
    - Commodities
    - Date
    - Low_price
    - High_Price
    - Sales Total
    - Total_Qty_Sold
    - Total_Kg_Sold
    - Stock_On_Hand
    - avg_price_per_kg
    

The types of visualisations we can create differ depending on the data type of each variable. Broadly, we need to determine whether a variable is numerical, or categorical. Currently the **categorical** features are `Province`, `Size_Grade`, and `Commodities`. `Weight_Kg`, `Date`, `Low_Price`, `High_Price`, `Sales_Total`, `Total_Qty_Sold`, `Total_Kg_Sold`,`Stock_On_Hand`  and `avg_price_per_kg` are **numerical**. This is confirmed by looking specifically at the data type of each feature.

In [7]:
dftrain.shape

(64376, 13)

The shape command shows us that we have **64376** rows of data and **13** features.

In [8]:
dftrain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64376 entries, 0 to 64375
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Province          64376 non-null  object 
 1   Container         64376 non-null  object 
 2   Size_Grade        64376 non-null  object 
 3   Weight_Kg         64376 non-null  float64
 4   Commodities       64376 non-null  object 
 5   Date              64376 non-null  object 
 6   Low_Price         64376 non-null  float64
 7   High_Price        64376 non-null  float64
 8   Sales_Total       64376 non-null  float64
 9   Total_Qty_Sold    64376 non-null  int64  
 10  Total_Kg_Sold     64376 non-null  float64
 11  Stock_On_Hand     64376 non-null  int64  
 12  avg_price_per_kg  64376 non-null  float64
dtypes: float64(6), int64(2), object(5)
memory usage: 6.4+ MB


The info command confirms our categorical and numerical features. If a feature (variable) is categorical the Dtype is object and if it is a numerical variable the Dtype is an int64 or float64. This command also shows us that out of the 64376 none of the features contain any null values.

Null values for each feature can also be checked by using the following command:

- df.isnull().sum()
This command will provide the total number of null values appearing in each feature.

In [9]:
dftrain.isnull().sum()

Province            0
Container           0
Size_Grade          0
Weight_Kg           0
Commodities         0
Date                0
Low_Price           0
High_Price          0
Sales_Total         0
Total_Qty_Sold      0
Total_Kg_Sold       0
Stock_On_Hand       0
avg_price_per_kg    0
dtype: int64

This confirms that there are no null values in the data.

### Univariate Analysis: Non-Graphical

The first univariate analysis will be non-graphical. This is where we will be looking at the **descriptive statistics** of each feature. We can get the descriptive statistics of each numerical feature by using the following command:

    - df.describe()

This command will provide the mean, standard deviation and a five number summary of each numerical feature.
The five number summary (Minimum, Lower Quartile (Q1) = 25%, Median (Q2) = 50%, Upper Quartile (Q3) = 75%, Maximum) is also used for creating the box plot.

Individual statistical measures can also be calculated by using the following commands:

    - df.count()
    - df.mean()
    - df.std()
    - df.min()
    - df.quantile([0.25, 0.5, 0.75], axis = 0)
    - df.median()
    - df.max()

The three measures for central tendency are the mode, mean and median. The command to determine the mode is:

    - df.mode()

In [10]:
dftrain.describe()

Weight_Kg     Low_Price    High_Price   Sales_Total  Total_Qty_Sold  \
count  64376.000000  64376.000000  64376.000000  6.437600e+04    64376.000000   
mean      12.781592     75.651938     89.607858  1.939501e+04      446.104402   
std       35.943052    159.508144    172.223177  4.442192e+04     1184.169758   
min        0.120000      1.000000      1.000000 -5.770000e+04     -595.000000   
25%        7.000000     30.000000     35.000000  1.154000e+03       20.000000   
50%       10.000000     46.000000     55.000000  5.400000e+03      107.000000   
75%       11.000000     80.000000    100.000000  1.877200e+04      390.000000   
max      500.000000   4400.000000   4400.000000  1.134701e+06    39453.000000   

       Total_Kg_Sold  Stock_On_Hand  avg_price_per_kg  
count   64376.000000   64376.000000          64376.00  
mean     3336.641295     477.646328               NaN  
std      7682.295441    1453.892091               NaN  
min     -5040.000000    -512.000000              -inf  
25%       175.000000       0.000000              4.02  
50%       940.000000      76.000000              6.00  
75%      3250.000000     381.000000              8.67  
max    192230.000000   93193.000000               inf

# Calculating Kurtosis and Skewness

In [11]:
dftrain.skew()

Weight_Kg           10.746241
Low_Price           10.071797
High_Price           9.519075
Sales_Total          7.372393
Total_Qty_Sold       9.448153
Total_Kg_Sold        7.171355
Stock_On_Hand       13.767932
avg_price_per_kg          NaN
dtype: float64

In [12]:
dftrain.kurtosis()

Weight_Kg           116.022033
Low_Price           123.977754
High_Price          109.763276
Sales_Total          89.378966
Total_Qty_Sold      152.646835
Total_Kg_Sold        83.222862
Stock_On_Hand       437.075702
avg_price_per_kg           NaN
dtype: float64

In [15]:
dftrain.sort_values(by='Weight_Kg', ascending=False).head(10)

Province Container Size_Grade  Weight_Kg           Commodities  \
11828       CAPE     IG500         1S      500.0  GRAPEFRUIT STAR RUBY   
54870      NATAL     IG500         1S      500.0      ORANGE VALENCIAS   
62985  TRANSVAAL     IG500         1M      500.0  GRAPEFRUIT STAR RUBY   
19448       CAPE     IG500         1L      500.0      ORANGE VALENCIAS   
40198       CAPE     IG500         2L      500.0  GRAPEFRUIT STAR RUBY   
63590       CAPE     IG500         1M      500.0                 LEMON   
29584       CAPE     IG500         3M      500.0  GRAPEFRUIT STAR RUBY   
19496      NATAL     IG500         1M      500.0  GRAPEFRUIT STAR RUBY   
15350  TRANSVAAL     IG500         2S      500.0          ORANGE OTHER   
53652      NATAL     IG500         1M      500.0  GRAPEFRUIT STAR RUBY   

             Date  Low_Price  High_Price  Sales_Total  Total_Qty_Sold  \
11828  2020-07-31     1200.0      1200.0       2400.0               2   
54870  2020-08-27      500.0       500.0       7000.0              14   
62985  2020-07-09      600.0       600.0      12000.0              20   
19448  2020-09-17      500.0       500.0       3000.0               6   
40198  2020-06-17     1000.0      1000.0      10000.0              10   
63590  2020-08-27     2000.0      2500.0      10500.0               5   
29584  2020-05-26     1000.0      1000.0       2000.0               2   
19496  2020-07-01      600.0       600.0      19200.0              32   
15350  2020-08-18     2000.0      2000.0       4000.0               2   
53652  2020-07-09      600.0       600.0      19200.0              32   

       Total_Kg_Sold  Stock_On_Hand  avg_price_per_kg  
11828         1000.0              0               2.4  
54870         7000.0              0               1.0  
62985        10000.0              0               1.2  
19448         3000.0              0               1.0  
40198         5000.0              0               2.0  
63590         2500.0             19               4.2  
29584         1000.0              0               2.0  
19496        16000.0              0               1.2  
15350         1000.0              0               4.0  
53652        16000.0              0               1.2

## Identify the target attributes.

The target attribute is the price per kilogram for the APPLE GOLDEN DELICIOUS Commodity

## Visualize the data.

## Study the correlations between attributes.

## Study how you would solve the problem manually.

## Identify the promising transformations you may want to apply.

## Identify the data that would be useful.

## Document what you have learned.

# Prepare the data

Notes:
* Work on copies of the data (keep the original dataset intact)
* Write functions for all data transformations you apply, for five reasons:
    * So you can easily prepare the data the next time you get a fresh dataset
    * So you can apply these transformations in future projects
    * To clean and prepare a test set
    * To clean and prepare new instances once your solution is live
    * To make it easy to treat your preparation choices as hyperparameters
    

## Data Cleaning:

* Fix or remove outliers (optional).
* Fill in missing values (e.g. with zero, mean, median, ...) or drop their rows/columns.

## Feature selection (optional):

* Drop the attributes that provide no useful information for the task.

## Feature engineering, where appropriate

* Discretize continuous features
* Decompose features (e.g. categorical, date/time, etc.)
* Add promising transformations of features (e.g. log(x), sqrt(x), x^2, etc.)
* Aggregate features into promising new features


## Short-listing Promising Models

Notes:
* If the data is huge, you may want to sample smaller training sets so you can train many different models in a reasonable time (be aware that this penalizes complex models such as large neural nets and Random Forests).
* Once again, try to automate these steps as much as possible

1. Train many quick and dirty models from different categories (e.g. linear, naive Bayes, SVM, Random Forests, neural net, etc.) using standard parameters.
2. Measure and compare their performance.
    * For eah model, use N-fold cross validation and compute mean and standard deviation of the performance measure of the N folds.
3. Analyze the most significant variables for each algorithm.
4. Analyze the types fo errors the models make.
    * What data would a human have used to avoid the errors?
5. Have a quick round of feature selection and engineering.
6. Have one or two more quick iterations of the five previous steps.
7. Short-list the top three to five most promising models, preferring the models that make different types of errors.


## Fine-Tune the system.

Notes:
* You will want to use as much data as possible for this step, especially as you move toward the end of fine-tuning.
* As always automate what you can.

1. Fine-tune the hyperparameters using cross-validation.
    * Treat your data transformation choices as hyperparameters, especially when you are not sure about them (e.g. should I replace missing values with zero or with the median value? Or just drop rows?).
    * Unless htere are very few hyperparameter values to explores, prefer random search over grid search. If training is very long, you may prefer a Bayesian optimization approach (e.g. using Gaussian process priors: https://arxiv.org/pdf/1206.2944.pdf)
2. Try ensemble methods. Combining your best models will often perform better than running them individually.
3. Once you are confident about your final model, measure its performance on the tes set to estimate the generalization error.

!
Don't tweak your model after measuring the generalization error: You would just start overfitting the test set.

# Present your solution

## Document what you have done.

## Create a nice presentation.

* Make sure you highlight the big picture first.

## Explain why your solution achieves the business objective.

## Don't forget to present interesting points you noticed along the way.

* Describe what worked and what did not.
* List your assumptions and your systems' limitations.

## Ensure your key findings are communicated through beautiful visualizations or easy to remember statements (e.g. "the median income is the number-one predictor of housing prices").

# Launch

## Get your solution ready for production (plug into production data inputs, write unit tests, etc.)

## Write monitoring code to check your systems' live performance at regular intervals and trigger alerts when it drops.

* Beware of slow degradation too: models tend to "rot" as data evolves.
* Measuring performance may require a human pipeline (e.g. via crowdsourcing service).
* Also monitor your inputs' quality (e.g. a malfunctioning sensor sending random values, or another team's output becoming stale). This is particularly important for online learning systems. 

## Retrain your models on a regular basis on fresh data (automate as much as possible).